Implement additional SSN validation for federally-funded TANF recipients

elipe17 opened the pull request:
## Summary of Changes
- Add function to TDR parser to handle new error requirement
- Add test for new function/error
- Add mixin so records know what line number they came from
- Migration for new field
- Updated all models
Pull request closes #5333

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit a TANF section 1 file that has at least one case that has a T1 record with FUNDING_STREAM=1 and a T2 record with FAMILIY_AFFILIATION=1 and an invalid SSN. e.g. 0001231234
OR
3. Submit the attached file and verify the error report lists the error as critical
[test.txt](https://github.com/user-attachments/files/22566024/test.txt)
<img width="1502" height="483" alt="Screenshot 2025-09-26 at 12 59 01 PM" src="https://github.com/user-attachments/assets/a5d5254c-b147-4a59-b9ed-bf933251329b" />


## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] A new validation rule is created for SSNs in TANF Section 1: Active Case Data  
- [ ] Rule applies to:  
  - T1, Item 9 = Funding Stream = 1 (federally-funded)  
  - T2, Items 30 and 67 = Family Affiliation = 1 (recipient)  
- [ ] Invalid SSNs are identified according to the rules listed above  
- [ ] Invalid SSNs trigger an error in the **critical tab** of the error report  
- [ ] Records with invalid SSNs continue processing unless flagged by other Category 1 or 4 validations  
- [ ] Testing Checklist has been executed and all tests pass  
- [ ] README and documentation updated if necessary  
- [ ] [Release Digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) updated with testing steps and media  

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Remove commented code
- Add function to TDR parser to handle new error requirement
- Add test for new function/error
- Add mixin so records know what line number they came from
- Migration for new field
- Updated all models

elipe17 committed:
- Update tests

elipe17 committed:
- linting

elipe17 committed:
- Update error report generator to make the new error critical
- Update the error generator function to have all the correct info for the error report
- use correct call from error factory

elipe17 committed:
- Update test doc string

elipe17 commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -367,3 +370,90 @@ def delete_serialized_records(self):
+                                    error_generator(generator_args=generator_args)
+                                ]
+                            }
+                        )
```
Call bulk create errors and increment error count

elipe17 committed:
- increment errors and call bulk create to ensure we dont overflow the errors cache

elipe17 commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -367,3 +370,92 @@ def delete_serialized_records(self):
+            category2.valueNotAt(slice(0, 3), "666"),
+            category2.valueNotAt(slice(3, 5), "00"),
+            category2.valueNotAt(slice(5, 9), "0000"),
+            error_message="Federally funded recipients must have a valid SSN.",  # TODO: need input on this
```
@reitermb Could I also get some UX input here for what the error message should be to correctly convey the error?

victoriaatraft replied:
@elipe17 would it be possible to call out what the error is specifically that makes it an invalid SSN?

elipe17 replied:
It is an invalid SSN if: it is not a number, is not a nine digit number, begins with three zeros, begins with three sixes, contains zeros in the fourth and fifth positions, or contains all zeros in the last four digits.

victoriaatraft replied:
I could have sworn we have similar error messages re SSN, but suggestion that comes to mind:
`SSN is invalid because it is {invalid error(s)}`

reitermb replied:
I think I might lean here towards message consistency with our existing invalid SSN error and maybe adding an enhancement as a follow-on to link out to a knowledge center mini-page (a la the FTANF one) breaking down SSN rules?

Being able to identify what's wrong would be cool, but I think the reality of what fixing these looks like is almost always going to be getting case workers to provide the correct data rather than actually doing any troubleshooting with the number given. Thoughts on that @victoriaatraft?

elipe17 replied:
The message we use for FRA invalid SSN is:
`Social Security Number is not valid. Check that the SSN is 9 digits, does not contain only zeroes in any one section, and does not contain dashes or other punctuation. Enter 999999999 if an individual does not have an SSN.`

elipe17 committed:
Merge branch 'develop' into 5333-ssn-ff-tanf-recips

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 5333-ssn-ff-tanf-recips

jtimpe commented on tdrs-backend/tdpservice/data_files/error_reports.py:
```diff
@@ -327,6 +327,7 @@ def get_prioritized_queryset(self):
             ("FAMILY_AFFILIATION", "EDUCATION_LEVEL"),
             ("FAMILY_AFFILIATION", "WORK_ELIGIBLE_INDICATOR"),
             ("CITIZENSHIP_STATUS", "WORK_ELIGIBLE_INDICATOR"),
+            ("FUNDING_STREAM", "FAMILY_AFFILIATION", "SSN"),
```
irrelevant for this pr, but since we're only interested in particular fields for this query, we could possibly represent it with `cat2_priority=True` and `cat3_priority=True` members on the `Field`. Then maybe `RowSchema` or `SchemaManager` helpers for obtaining a list and/or constructing a query. feels more declarative in the "provide all the options with the schema definition" sense. just thought i'd write that down somewhere

raftmsohani replied:
I like Jan's suggestion

jtimpe commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -358,3 +363,92 @@ def _generate_trailer_errors(self, trailer_errors):
+                    result = validator(ssn, eargs)
+                    if not result.valid:
+                        error_generator = self.error_generator_factory.get_generator(
+                            ErrorGeneratorType.VALUE_CONSISTENCY, t2_record.line_number
```
these aren't necessarily the same error type as case consistency, but i think the idea is similar: validate one record's values (T2 SSN) based on another record's (T1 Funding Stream). i'm thinking about how to incorporate this in a more generic way, especially if we have to take this functionality over to the `ProgramAuditParser` as well

elipe17 replied:
Yes absolutely. What I want (or at least think will help) would be to treat all parsing and validation in the same way we do with cat4. That is, parse all records in a case first, validate after we have the whole case. Then, we could also perform hierarchical validation. E.g validate cat1 first, if we have cat1 errors delete the record and therefore the case since it will generate cat4 errors. Then validate cat4, and so on. That way we could avoid generating errors for records/cases that are going to be thrown away anyways. 

If we accomplished this, we could have much looser rules of what cat1 - cat4 errors and validation rules are. This format would also set us up nicely to begin processing cases and or sets of records in parallel.

jtimpe replied:
i like the sound of that!

jtimpe commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -358,3 +363,92 @@ def _generate_trailer_errors(self, trailer_errors):
+        t1_fields = [t1_schema.get_field_by_name(name) for name in t1_field_names]
+        t2_fields = [t2_schema.get_field_by_name(name) for name in t2_field_names]
+        fields = t1_fields + t2_fields
+        while True:
```
do we need `while True` since we aren't deleting records from the queryset as we process?

elipe17 replied:
Good catch. I don't think we do, but let me double check

jtimpe reviewed (commented):
Not quite sure what's happening or i'd point to the code, but it seems like only the _last_ SSN error in the file is being reported. In `small_tanf_section1.txt`, I updated/verified that several T2 records amongst several cases are invalid. I wasn't sure why they weren't showing up in the spreadsheet, until i started deleting cases further down the list and the errors for the previous cases started to show. That appears to be the case for the Summary and Critical tabs.

elipe17 commented:
> Not quite sure what's happening or i'd point to the code, but it seems like only the _last_ SSN error in the file is being reported. In `small_tanf_section1.txt`, I updated/verified that several T2 records amongst several cases are invalid. I wasn't sure why they weren't showing up in the spreadsheet, until i started deleting cases further down the list and the errors for the previous cases started to show. That appears to be the case for the Summary and Critical tabs.

I will take a look and see what I find out!

elipe17 committed:
- Fix indexing error causing error creatio to be skipped
- Updated error message

elipe17 commented:
> > Not quite sure what's happening or i'd point to the code, but it seems like only the _last_ SSN error in the file is being reported. In `small_tanf_section1.txt`, I updated/verified that several T2 records amongst several cases are invalid. I wasn't sure why they weren't showing up in the spreadsheet, until i started deleting cases further down the list and the errors for the previous cases started to show. That appears to be the case for the Summary and Critical tabs.
> 
> I will take a look and see what I find out!

@jtimpe I found the issue. I made a mistake with my index incrementing logic. The index was being incremented in the `for` loop instead of the `while` loop. I removed the `while` loop and leveraged the queryset iterator instead. That fixed the missing errors you were seeing. Good catch on that!

elipe17 committed:
Merge branch 'develop' into 5333-ssn-ff-tanf-recips

elipe17 committed:
- Update tests

elipe17 committed:
Merge branch '5333-ssn-ff-tanf-recips' of https://github.com/raft-tech/TANF-app into 5333-ssn-ff-tanf-recips

raftmsohani commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -358,3 +363,88 @@ def _generate_trailer_errors(self, trailer_errors):
+        validator = category2.ssnAllOf(
+            category2.isNumber(),
+            category2.intHasLength(9),
+            category2.valueNotAt(slice(0, 3), "000"),
```
nit: can we add the constants to the SSN as a constant to schema defs? in the line of: SSN_INVALID=["000", ... or something like that. I think we should centralize the configs as much as possible in the schema

elipe17 replied:
Are you suggesting define the `000` as a constant in the schema def and then import it into the parser?

raftmsohani replied:
yes

elipe17 replied:
Take a look at what I did and let me know what you think!

raftmsohani commented on tdrs-backend/tdpservice/parsers/tdr_parser.py:
```diff
@@ -189,6 +190,10 @@ def parse_and_validate(self):
         )
         self.case_consistency_validator.clear_errors()
 
+        self.generate_funded_ssn_errors()
```
nit: not related but: we are growing the parse_and_validate method. We should create Logical steps and dismantle this function like:
1) Initialize or setup validators, etc
2) Evaluate header
3) evaluate trailer
4) Schema check
...

These are obviously suggestion, but we need to clearly define the funtional steps at a bird-eye level

jtimpe reviewed (approved):
Looks good!

elipe17 committed:
- Created constants file
- Updated schema and parser class to leverage constants

elipe17 committed:
- Linting
